lib/resty/kafka/protocol/record.lua

```diff
@@ -0,0 +1,198 @@
+local protocol = require("resty.kafka.protocol.common")
+local bit = require("bit")
+local math = require("math")
+
+local crc32c = protocol.crc32c
+local ngx_log = ngx.log
+local ngx_crc32 = ngx.crc32_long
+local ERR = ngx.ERR
+local table_insert = table.insert
+local band = bit.band
+local math_abs = math.abs
+
+
+local _M = {}
+
+
+--- Decode MessageSet v0 and v1 (the old format) data.
+-- Tip: The return value message contains the int64 value, which is of
+-- type cdata and cannot be used directly in some scenarios.
+-- @author bzp2010 <bzp2010@apache.org>
+local function _message_set_v0_1_decode(resp)
+    local message = {}
+
+    message.offset = resp:int64()
+
+    -- Sometimes in low version protocols many useless empty data are
+    -- returned, they are characterized by a -1 offset at the beginning
+    -- and 0 for each offset after that, we will discard the data after that.
+    -- The response will be closed (offset is set to the end),
+    -- i.e. it cannot continue reading any data.
+    -- Tip: offset uses int64 storage, and it is almost impossible for us
+    -- to write so many messages, so the case where offset does have a
+    -- negative value is not considered here.
+    if message.offset < 0 then
+        resp:close()
+        return {} -- return an empty message table
+    end
+
+    local message_size = resp:int32()
+    local crc = resp:int32()
+    local crc_content = resp:peek_bytes(resp.offset, message_size - 4)
+    local calc_crc = ngx_crc32(crc_content)
+    if crc ~= calc_crc and math_abs(crc) + math_abs(calc_crc) ~= 4294967296 then
+        return nil, "crc checksum error"
+    end
+
+    local magic_byte = resp:int8()
+
+    -- TODO: support compressed Message Set
+    local attributes = resp:int8()
+
+    -- message version 1 added timestamp
+    if magic_byte == protocol.API_VERSION_V1 then
+        message.timestamp = resp:int64()
+    else
+        message.timestamp = 0
+    end
+
+    message.key = resp:bytes()
+    message.value = resp:bytes()
+
+    return {message}, nil
```

Comment:
let us remove `, nil`, it is useless

less is better

Reply:
removed